### ISSUE TITLE
angular-translate 3.0.0

### ISSUE THREAD

**Issue by knalli** [angular-translate/angular-translate#995]
We should talk about this :)

In general, this is not a new thing, however, it was only spoken and written down anywhere. We should plan a next major release, perhaps not with many new features. But with the opportunity to cut down some "legacy strings".

* [ ] Dropping support for AJS 1.2, updating internal base to AJS 1.3.x latest
  * [ ] `filter.$stateful` management is not optional anymore (if-branch can be omitted)
  * [ ] with a re-arrangement of the escaping strategies we can use `ngSanitize`
* [ ] better security by default (latest mention in #993), already announced in [docs](http://angular-translate.github.io/docs/#/guide/19_security))
* [ ] re-arrangement of modules
 * [ ] discussion state of satellite bower shadow repositories -- we should discuss whether we need them or we can combine them all files into one (dedicated) repository (reasons pro/contra?). 
 * [ ] re-think the current internal module dependency hierarchy -- do we need dedicated modules actually? Why? We do not need them for a plugin architecture.
 * [ ] a full working migration guide (more or less a well written list of the breaking changes)

This is community driven, so we need everyone's input, additions or comments.

/cc @angular-translate/core @angular-translate/members

**Comment by DWand** [comment]
@knalli, and how about synchronization of processes which are taking place in `use` and `refresh` methods? It looks like a big fundamental piece of work which affects internal architecture of the service.

**Comment by knalli** [comment]
What do you mean in detail?

**Comment by DWand** [comment]
@knalli, I'm not sure if this issue is still valid, but most likely it is. I've described it in the issue #449. As I remember, there were also a few other issues somehow related to the question of synchronization.

**Comment by DWand** [comment]
I disagree with at least 2 options:

1) I think that it is not good to make **all** things asynchronous. Why should something pretend to be asynchronous if it is synchronous by nature? I think that such tools have to be used only if they are needed to accomplish the task. Needless promises may confuse the developer and lead to misunderstanding of how the thing works.

2) The `activeLanguage` method looks like a getter. If we try to make things more clear and easy to understand, then it might be better to split the `use` method to `getActiveLanguage` and `setActiveLanguage`. This is Java style, but people use it for years and it works pretty well.

**Comment by knalli** [comment]
1.  More the promise style allowing a deferred setup. And if A is internally deferred, than B will it be also automatically. 
2.  My idea was to stick with the current way of a chaining configuration.

**Comment by knalli** [comment]
Addition notes on "re-arrangement of modules"

* Looking at loader-staticfiles and loader-url, I don't see any cause why this should be a dedicated plugin anymore. Neither it is a lot of code (even static files is not so much), nor it depends on a non AJS core component. Merging both loaders into one file loader combined with with some intelligent convenient method should be a lot more helpful than the current situation. Speaking for modularization, we merge them at least into one xhr loader plugin.
* Looking at handler-log, I would replace the pattern with a configuration (just like customerLoader works). This oneliner plugin could be discarded.

And regarding dependencies:
* storage-cookie can get a proper dependency on ngCookie
* if we make the storage api promisable and "null"-aware, we can eventuelly get rid of the cookies fallback. However, an alternative fallback strategy should be available.

**Comment by DWand** [comment]
Also, it might be an interesting addition to create some king of constructor - a piece of software to create personal builds of the module. This way the user will be able to pack all needed related modules to one js file. Probably it may also reduce amount of duplicate code in plugin somehow. But it looks tricky and is hardly needed to anyone.

**Comment by knalli** [comment]
Well, that syntactical sugar is only recipient for custom bower/npm/browerify solution. The idea isn't bad, but should not be on primary focus. (imho)

**Comment by knalli** [comment]
Note. I'll try to start final concepts in the first week of June, just after a short run for 2.7.1 incl. some fixes and 2.7 additions.

**Comment by knalli** [comment]
@PascalPrecht @tspaeth 

* Currently, the module name is `pascalprecht.translate` with the domain `pascalprecht`. Is this still valid now? But which one would be the appropriate one now?
* Years ago I had read plugin's services should not prefixed with `$` (as we did in `$translate`). Is this still valid? In this case, should it be `translate` only?

**Comment by PascalPrecht** [comment]
I'd vote for keeping namespaces and service names as they are. People got used to them and angular-translate is very widely used in a lot of big applications. This release would introduce a huge breaking change then. I think it's okay to just stick with that, other community driven projects like ui-router also use `$` prefix in their services.

My 0.02$

**Comment by knalli** [comment]
Ok, good points. Probably the breaking issue could be handled with a compat-module (planned already). I only want to take the chance now. Last time.. ;)

**Comment by tspaeth** [comment]
Regarding the service names - I don't have a 100% strong opinion. Usually on the application architectures we build, we try to avoid $ as prefix for our application and its modules to differentiate it from "base modules delivered by the framework" or similar. So from the other perspective: It is okay as a base module like this for me to have it prefixed.

On the namespace whilst keeping respect to Pascal, an organizational name as module base from a (whenever it might exist) next major version with a compat way of being partly backwards compatible sounds also ok to me. As the refactoring effort to me is somewhat high by looking at the checklist, I see anyway some issues on keeping a 100% compatibility with existing applications and their loaders / load mechanisms. (see async, fallback, filter issues etc.) 
My 0.0199 Pence ;-)

**Comment by knalli** [comment]
Speaking of a [compat service](https://gist.github.com/knalli/84a58e305f4f7904cbdf). I have done a lot of the work already yesterday, but still many points of the checklist are todo. I'll probably will create a temporarily repository until we decide the migration.

If we stick with the prefix, the compat service will/must be a decorator, I think.

**Comment by knalli** [comment]
As already mentioned earlier this day, I've pushed the first stuff into a temporarily [dedicated repository](https://github.com/angular-translate/angular-translate-v3) (look into `src2/`). If you are with me, I'll rebase the stuff with proper messages and push it into the regular repository (in branch canary-v3 or so).

What does contain this preview?

* the usage of the module (in demo [ex0](https://github.com/angular-translate/angular-translate-v3/blob/71eb27c79cabc3806140a2961b8196d3c788e555/demo/ex0.htm#L68-L144)). Very similar.
* service name is `translate`, but we have already discussion to use the old name `$translate`
* no final decision: `$translate.instant` => `translate.sync` (as seen in [compat](https://github.com/angular-translate/angular-translate-v3/blob/71eb27c79cabc3806140a2961b8196d3c788e555/src2/compat/compat.module.js#L24-L26)
* first approach of a [compat module](https://github.com/angular-translate/angular-translate-v3/blob/71eb27c79cabc3806140a2961b8196d3c788e555/src2/compat/compat.module.js) for an easier migration (i.e. logging, auto converting calls)
* internals everything is working with multiple requests (not translation by id, but translations by ids)
* internal everything is working with a request and response metadata object (allowing additional information to transport for a translation)
* the [structure overview](https://gist.github.com/knalli/4871851b53506d2ea88b)
  * everything is ES6 (tooling: Browserify + BabelJS) => atm invoke `grunt v3`
  * everything is a module
  * only the AJS specific components composing the final directives and services: that is allowing us additional features like [directive aliasing](https://github.com/angular-translate/angular-translate-v3/blob/71eb27c79cabc3806140a2961b8196d3c788e555/src2/components/components.module.js#L20-L21).

What does NOT contain this preview?

* tests are completely untouched
  * my idea: try to enable as much as possible of the old tests with compat mode => legacy integration test
  * invent new unit tests (real unit tests, w/o angular)
  * invent new integration / "composed unit" tests (probably based on the old ones)
* features: storage, cache, partial loader, resolve locale/language, fallback, interpolation params, full feature directives (well: you will see the basics are working in demo ex0)
* documentation (a lot can be migrated)

So, I need feedback: You see the chosen architecture and way of composing everything. OK? Not OK? Comments? Perhaps someone have a fundamental remark on design?

**Comment by fracz** [comment]
I would leave the `filter.$stateful` configuration and conditional for applications that [do not need to change the language on the fly feature](https://github.com/angular-translate/angular-translate/issues/945). Stateful filters are [generally not recommended](https://docs.angularjs.org/guide/filter#stateful-filters) as they are evaluated on every digest cycle. Changing the language with "normal" link and page refresh seems to be a better idea. User generally does not change the language very often...

**Comment by knalli** [comment]
@fracz Yeah, I'd already the same idea because the same reasons. I've updated the list.

**Comment by RichardBradley** [comment]
Consider changing the error info when the Promise returned by `$translate` fails. See discussion on #1516. 
Using the translation key as the error info is a) unconventional, as this param is usually information about the error, b) redundant, as the caller will already have the translation key to hand, and c) ambiguous as it prevents any other error info being passed here.
